File: ios/DesMoinesInsider/ViewModels/EventsViewModel.swift
import Foundation
import CoreLocation

/// ViewModel for the home/events feed. Handles fetching, filtering, and pagination.
@MainActor
@Observable
final class EventsViewModel {
    // MARK: - State

    private(set) var events: [Event] = [] {
        didSet { recomputeArrangedEvents() }
    }
    /// Sponsored-arranged view of `events`, cached so the arrangement
    /// (SponsoredArranger.arrange) runs only when `events` changes — not on
    /// every SwiftUI body pass (IOS-AUDIT-PERF-010).
    private(set) var arrangedEvents: [Event] = []
    private(set) var featuredEvents: [Event] = []
    private(set) var isLoading = false
    private(set) var isLoadingMore = false
    private(set) var hasMore = false
    private(set) var totalCount = 0
    private(set) var errorMessage: String?

    // MARK: - Filters

    var searchText = "" {
        didSet { if oldValue != searchText { resetAndFetch() } }
    }
    var selectedCategory: EventCategory? {
        didSet { if oldValue != selectedCategory { resetAndFetch() } }
    }
    var selectedDatePreset: DateFilterPreset? {
        didSet { if oldValue != selectedDatePreset { resetAndFetch() } }
    }
    var showFeaturedOnly = false {
        didSet { if oldValue != showFeaturedOnly { resetAndFetch() } }
    }
    var selectedCities: Set<String> = [] {
        didSet { if oldValue != selectedCities { resetAndFetch() } }
    }
    /// Sort order for the events list. IOS-DISCOVER-2026-003.
    var sortBy: EventSortOption = .soonest {
        didSet { if oldValue != sortBy { resetAndFetch() } }
    }

    /// Currently applied smart preset, if any. Cleared when the user changes
    /// any individual filter manually.
    var activePreset: EventPreset? = nil

    // Premium filters (Insider+ only)
    var showFreeOnly = false {
        didSet { if oldValue != showFreeOnly { resetAndFetch() } }
    }
    var maxDistance: Double? {
        didSet { if oldValue != maxDistance { resetAndFetch() } }
    }
    var minRating: Double? {
        didSet { if oldValue != minRating { resetAndFetch() } }
    }

    /// Set while a bulk update (`clearFilters` / `applyPreset`) mutates many
    /// filter properties at once, so each individual `didSet` skips its own
    /// debounced fetch. The bulk operation fires exactly one `resetAndFetch()`
    /// at the end (IOS-AUDIT-PERF-004).
    private var isBulkUpdating = false

    // MARK: - Pagination

    private var currentOffset = 0
    /// Count of RAW server rows fetched so far. Pagination offset is driven by
    /// this — not `events.count` — so the client-side premium maxDistance filter
    /// dropping rows can't desync the server offset (repeat/skip events).
    private var rawFetchedCount = 0
    private let pageSize = Config.defaultPageSize
    private var fetchTask: Task<Void, Never>?

    private let service = EventsService.shared
    private let cache = QueryCache.shared

    // MARK: - Initial Load

    func loadInitialData() async {
        guard events.isEmpty else { return }

        // Serve cached data immediately for instant cold start
        let cacheKey = eventsCacheKey()
        let isOffline = !NetworkMonitor.shared.isConnected

        if let cached: [Event] = await cache.get(cacheKey, allowStale: isOffline) {
            events = applyPremiumFilters(cached)
            isLoading = false
        }

        if let cachedFeatured: [Event] = await cache.get("featured-events", allowStale: isOffline) {
            featuredEvents = cachedFeatured
        }

        // Then fetch fresh data in the background (skip if offline and we have cache)
        if isOffline && !events.isEmpty { return }

        await fetchEvents(reset: true)
        await fetchFeaturedEvents()
    }

    func refresh() async {
        // Pull-to-refresh always bypasses cache
        await fetchEvents(reset: true)
        await fetchFeaturedEvents()
    }

    // MARK: - Fetch Events

    func fetchEvents(reset: Bool = false) async {
        if reset {
            currentOffset = 0
            rawFetchedCount = 0
            // Only show spinner if we have no cached data
            if events.isEmpty { isLoading = true }
        } else {
            isLoadingMore = true
        }
        errorMessage = nil

        do {
            var query = EventsService.EventsQuery()
            query.searchText = searchText.isEmpty ? nil : searchText
            query.category = selectedCategory?.rawValue
            query.isFeatured = showFeaturedOnly ? true : nil
            query.cities = selectedCities.isEmpty ? nil : Array(selectedCities)
            query.freeOnly = showFreeOnly
            query.sortBy = sortBy
            query.limit = pageSize
            query.offset = currentOffset

            if let preset = selectedDatePreset {
                let range = preset.dateRange
                query.dateStart = range.start
                query.dateEnd = range.end
            }

            let response = try await service.fetchEvents(query: query)
            guard !Task.isCancelled else { return }

            // Apply premium filters client-side
            let filtered = applyPremiumFilters(response.events)

            if reset {
                events = filtered
                // Cache the first page for offline/cold-start use
                let cacheKey = eventsCacheKey()
                await cache.set(cacheKey, value: response.events)
                // Keep Spotlight in sync with what the user is browsing so events
                // are discoverable in system search (IOS-AUDIT-FEAT-026).
                let toIndex = response.events
                Task { await SpotlightService.shared.indexEvents(toIndex) }
            } else {
                events.append(contentsOf: filtered)
            }

            totalCount = response.totalCount
            hasMore = response.hasMore
            // Advance the offset by RAW rows fetched, not the filtered display
            // count, so the premium maxDistance filter can't shift the window.
            rawFetchedCount += response.events.count
            currentOffset = rawFetchedCount
        } catch {
            // If offline and we have cached data, don't overwrite with an error
            if events.isEmpty {
                errorMessage = error.localizedDescription
            }
        }

        isLoading = false
        isLoadingMore = false
    }

    // MARK: - Load More

    func loadMoreIfNeeded(currentItem: Event?) async {
        guard let currentItem,
              hasMore,
              !isLoadingMore,
              let index = events.firstIndex(where: { $0.id == currentItem.id }),
              index >= events.count - 5 else { return }

        await fetchEvents(reset: false)
    }

    // MARK: - Featured Events

    private func fetchFeaturedEvents() async {
        do {
            let featured = try await service.fetchFeaturedEvents()
            featuredEvents = featured
            await cache.set("featured-events", value: featured)
        } catch {
            // Keep existing cached featured events on failure
            if featuredEvents.isEmpty {
                featuredEvents = []
            }
        }
    }

    // MARK: - Cache Key

    private func eventsCacheKey() -> String {
        var parts = ["events"]
        if let cat = selectedCategory { parts.append("cat-\(cat.rawValue)") }
        if let preset = selectedDatePreset { parts.append("date-\(preset.rawValue)") }
        if showFeaturedOnly { parts.append("featured") }
        if !searchText.isEmpty { parts.append("q-\(searchText)") }
        parts.append("s-\(sortBy.rawValue)")
        return parts.joined(separator: "-")
    }

    // MARK: - Filter Management

    var activeFilterCount: Int {
        var count = 0
        if selectedCategory != nil { count += 1 }
        if selectedDatePreset != nil { count += 1 }
        if showFeaturedOnly { count += 1 }
        if !searchText.isEmpty { count += 1 }
        if showFreeOnly { count += 1 }
        if !selectedCities.isEmpty { count += 1 }
        if maxDistance != nil { count += 1 }
        if minRating != nil { count += 1 }
        return count
    }

    func clearFilters() {
        // Coalesce the many property mutations into a single debounced fetch
        // (IOS-AUDIT-PERF-004).
        isBulkUpdating = true
        selectedCategory = nil
        selectedDatePreset = nil
        showFeaturedOnly = false
        searchText = ""
        showFreeOnly = false
        selectedCities = []
        maxDistance = nil
        minRating = nil
        activePreset = nil
        isBulkUpdating = false
        resetAndFetch()
    }

    // MARK: - Smart Presets

    /// Applies a bundled set of event filters in one tap. Tapping the same
    /// preset again clears all filters.
    func applyPreset(_ preset: EventPreset) {
        if activePreset == preset {
            clearFilters()
            return
        }
        // Coalesce the bundled mutations into a single debounced fetch
        // (IOS-AUDIT-PERF-004).
        isBulkUpdating = true
        selectedCategory = preset.category
        selectedDatePreset = preset.datePreset
        showFeaturedOnly = preset.featured
        showFreeOnly = preset.free
        selectedCities = []
        maxDistance = nil
        minRating = nil
        searchText = ""
        activePreset = preset
        isBulkUpdating = false
        resetAndFetch()
    }

    // MARK: - Sponsored Arrangement (IOS-AUDIT-PERF-010)

    /// Recomputes `arrangedEvents` off the render path whenever `events`
    /// changes. Sponsored listings are pulled to the front; organic order is
    /// otherwise preserved.
    private func recomputeArrangedEvents() {
        arrangedEvents = SponsoredArranger.arrange(events, isSponsored: { $0.isActivelySponsored })
    }

    // MARK: - Premium Filters (applied client-side)

    private func applyPremiumFilters(_ events: [Event]) -> [Event] {
        var result = events

        // `showFreeOnly` is now server-side; keep maxDistance + minRating client-side

        if let maxDistance, let userLocation = LocationService.shared.userLocation {
            result = result.filter { event in
                guard let coord = event.coordinate else { return true }
                let eventLocation = CLLocation(latitude: coord.latitude, longitude: coord.longitude)
                let distanceMiles = userLocation.distance(from: eventLocation) / 1609.34
                return distanceMiles <= maxDistance
            }
        }

        // minRating is available for future use when events have ratings
        return result
    }

    // MARK: - Debounced Search

    private func resetAndFetch() {
        // Bulk updates (clearFilters / applyPreset) mutate many properties in
        // one go; suppress their per-property fetches and fire one at the end.
        guard !isBulkUpdating else { return }
        fetchTask?.cancel()
        fetchTask = Task {
            try? await Task.sleep(for: .milliseconds(300))
            guard !Task.isCancelled else { return }
            await fetchEvents(reset: true)
        }
    }
}

File: android/app/src/main/java/com/desmoines/aipulse/ui/screens/home/EventsViewModel.kt
package com.desmoines.aipulse.ui.screens.home

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.desmoines.aipulse.data.model.DateFilterPreset
import com.desmoines.aipulse.data.model.Event
import com.desmoines.aipulse.data.model.EventCategory
import com.desmoines.aipulse.data.model.Restaurant
import com.desmoines.aipulse.data.model.SubscriptionTier
import com.desmoines.aipulse.data.remote.EventsQuery
import com.desmoines.aipulse.data.model.RestaurantSortOption
import com.desmoines.aipulse.data.remote.RestaurantsQuery
import com.desmoines.aipulse.data.repository.EventsRepository
import com.desmoines.aipulse.data.repository.RestaurantsRepository
import com.desmoines.aipulse.util.Config
import com.desmoines.aipulse.util.LocationService
import com.desmoines.aipulse.util.NetworkMonitor
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.format.DateTimeFormatter
import javax.inject.Inject

private const val TAG = "EventsViewModel"

/**
 * ViewModel for the Home screen / events feed.
 * Mirrors iOS EventsViewModel.swift — handles fetching, filtering, pagination, caching.
 */
@HiltViewModel
class EventsViewModel @Inject constructor(
    private val eventsRepository: EventsRepository,
    private val restaurantsRepository: RestaurantsRepository,
    private val networkMonitor: NetworkMonitor,
    private val locationService: LocationService,
) : ViewModel() {

    // region State

    private val _events = MutableStateFlow<List<Event>>(emptyList())
    val events: StateFlow<List<Event>> = _events.asStateFlow()

    private val _featuredEvents = MutableStateFlow<List<Event>>(emptyList())
    val featuredEvents: StateFlow<List<Event>> = _featuredEvents.asStateFlow()

    private val _restaurants = MutableStateFlow<List<Restaurant>>(emptyList())
    val restaurants: StateFlow<List<Restaurant>> = _restaurants.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isLoadingMore = MutableStateFlow(false)
    val isLoadingMore: StateFlow<Boolean> = _isLoadingMore.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _hasMore = MutableStateFlow(false)
    val hasMore: StateFlow<Boolean> = _hasMore.asStateFlow()

    private val _totalCount = MutableStateFlow(0)
    val totalCount: StateFlow<Int> = _totalCount.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // endregion

    // region Filters

    private val _searchText = MutableStateFlow("")
    val searchText: StateFlow<String> = _searchText.asStateFlow()

    private val _selectedCategory = MutableStateFlow<EventCategory?>(null)
    val selectedCategory: StateFlow<EventCategory?> = _selectedCategory.asStateFlow()

    private val _selectedDatePreset = MutableStateFlow<DateFilterPreset?>(null)
    val selectedDatePreset: StateFlow<DateFilterPreset?> = _selectedDatePreset.asStateFlow()

    private val _showFeaturedOnly = MutableStateFlow(false)
    val showFeaturedOnly: StateFlow<Boolean> = _showFeaturedOnly.asStateFlow()

    // Premium filters (Insider+ only)
    private val _showFreeOnly = MutableStateFlow(false)
    val showFreeOnly: StateFlow<Boolean> = _showFreeOnly.asStateFlow()

    private val _maxDistance = MutableStateFlow<Double?>(null)
    val maxDistance: StateFlow<Double?> = _maxDistance.asStateFlow()

    private val _minRating = MutableStateFlow<Double?>(null)
    val minRating: StateFlow<Double?> = _minRating.asStateFlow()

    // Current subscription tier (set externally when subscription state changes)
    private val _currentTier = MutableStateFlow(SubscriptionTier.FREE)
    val currentTier: StateFlow<SubscriptionTier> = _currentTier.asStateFlow()

    // endregion

    // region Derived State

    /**
     * Combined UI state for HomeScreen. Mirrors iOS approach of computed properties
     * exposed to the view, but uses Compose-friendly StateFlow.
     */
    @Suppress("UNCHECKED_CAST")
    val uiState: StateFlow<HomeScreenState> = combine(
        _events, _featuredEvents, _restaurants,
        _isLoading, _isLoadingMore, _isRefreshing,
        _hasMore, _totalCount, _errorMessage,
    ) { values: Array<Any?> ->
        HomeScreenState(
            events = values[0] as List<Event>,
            featuredEvents = values[1] as List<Event>,
            restaurants = values[2] as List<Restaurant>,
            isLoading = values[3] as Boolean,
            isLoadingMore = values[4] as Boolean,
            isRefreshing = values[5] as Boolean,
            hasMore = values[6] as Boolean,
            totalCount = values[7] as Int,
            errorMessage = values[8] as String?,
        )
    }.combine(
        combine(
            _selectedCategory, _selectedDatePreset, _showFeaturedOnly,
            _searchText, _showFreeOnly, _maxDistance, _minRating, _currentTier,
            networkMonitor.isConnected,
        ) { filterValues: Array<Any?> ->
            FilterState(
                selectedCategory = filterValues[0] as EventCategory?,
                selectedDatePreset = filterValues[1] as DateFilterPreset?,
                showFeaturedOnly = filterValues[2] as Boolean,
                searchText = filterValues[3] as String,
                showFreeOnly = filterValues[4] as Boolean,
                maxDistance = filterValues[5] as Double?,
                minRating = filterValues[6] as Double?,
                currentTier = filterValues[7] as SubscriptionTier,
                isConnected = filterValues[8] as Boolean,
            )
        }
    ) { state, filters ->
        state.copy(
            selectedCategory = filters.selectedCategory,
            selectedDatePreset = filters.selectedDatePreset,
            showFeaturedOnly = filters.showFeaturedOnly,
            activeFilterCount = computeActiveFilterCount(filters),
            currentTier = filters.currentTier,
            isConnected = filters.isConnected,
        )
    }.stateIn(
        scope = viewModelScope,
        started = SharingStarted.WhileSubscribed(5_000),
        initialValue = HomeScreenState(),
    )

    // endregion

    // region Pagination

    private var currentOffset = 0
    private val pageSize = Config.DEFAULT_PAGE_SIZE
    private var debounceJob: Job? = null
    private var hasLoadedInitialData = false

    // endregion

    // region Public API

    /**
     * Load initial data — serve cached data immediately, fetch fresh in background.
     * Skip network if offline with cache hit. Mirrors iOS loadInitialData().
     */
    fun loadInitialData() {
        if (hasLoadedInitialData) return
        hasLoadedInitialData = true

        viewModelScope.launch {
            _isLoading.value = _events.value.isEmpty()

            // Fetch events and featured events in parallel
            launch { fetchEvents(reset = true) }
            launch { fetchFeaturedEvents() }
            launch { fetchPopularRestaurants() }
        }
    }

    /**
     * Pull-to-refresh — always bypasses cache. Mirrors iOS refresh().
     */
    fun refresh() {
        viewModelScope.launch {
            _isRefreshing.value = true
            _errorMessage.value = null

            launch { fetchEvents(reset = true) }
            launch { fetchFeaturedEvents() }
            launch { fetchPopularRestaurants() }

            // Wait a beat then clear refreshing (individual fetches handle their own state)
            delay(300)
            _isRefreshing.value = false
        }
    }

    fun setSearchText(text: String) {
        _searchText.value = text
        resetAndFetch()
    }

    fun setSelectedCategory(category: EventCategory?) {
        _selectedCategory.value = category
        resetAndFetch()
    }

    fun setSelectedDatePreset(preset: DateFilterPreset?) {
        _selectedDatePreset.value = preset
        resetAndFetch()
    }

    fun setShowFeaturedOnly(featured: Boolean) {
        _showFeaturedOnly.value = featured
        resetAndFetch()
    }

    fun setShowFreeOnly(freeOnly: Boolean) {
        _showFreeOnly.value = freeOnly
        resetAndFetch()
    }

    fun setMaxDistance(distance: Double?) {
        _maxDistance.value = distance
        resetAndFetch()
    }

    fun setMinRating(rating: Double?) {
        _minRating.value = rating
        resetAndFetch()
    }

    fun setCurrentTier(tier: SubscriptionTier) {
        _currentTier.value = tier
    }

    fun clearFilters() {
        _selectedCategory.value = null
        _selectedDatePreset.value = null
        _showFeaturedOnly.value = false
        _searchText.value = ""
        _showFreeOnly.value = false
        _maxDistance.value = null
        _minRating.value = null
        resetAndFetch()
    }

    /**
     * Trigger load more when user scrolls near end of list.
     * Mirrors iOS loadMoreIfNeeded(currentItem:).
     */
    fun loadMoreIfNeeded(currentIndex: Int) {
        val events = _events.value
        if (!_hasMore.value || _isLoadingMore.value) return
        if (currentIndex < events.size - 5) return

        viewModelScope.launch {
            fetchEvents(reset = false)
        }
    }

    // endregion

    // region Private Implementation

    /**
     * Fetch events from repository. Mirrors iOS fetchEvents(reset:).
     */
    private suspend fun fetchEvents(reset: Boolean) {
        if (reset) {
            currentOffset = 0
            if (_events.value.isEmpty()) {
                _isLoading.value = true
            }
        } else {
            _isLoadingMore.value = true
        }
        _errorMessage.value = null

        val dateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE_TIME

        val query = EventsQuery(
            searchText = _searchText.value.ifEmpty { null },
            category = _selectedCategory.value?.displayName,
            dateStart = _selectedDatePreset.value?.dateRange?.first?.format(dateTimeFormatter),
            dateEnd = _selectedDatePreset.value?.dateRange?.second?.format(dateTimeFormatter),
            isFeatured = if (_showFeaturedOnly.value) true else null,
            limit = pageSize,
            offset = currentOffset,
        )

        eventsRepository.fetchEvents(query)
            .onSuccess { response ->
                val filtered = applyPremiumFilters(response.events)
                if (reset) {
                    _events.value = filtered
                } else {
                    _events.update { current -> current + filtered }
                }
                _totalCount.value = response.totalCount
                _hasMore.value = response.hasMore
                currentOffset = _events.value.size
            }
            .onFailure { error ->
                Log.e(TAG, "fetchEvents failed: ${error.message}", error)
                // If offline and we have cached data, don't overwrite with an error
                if (_events.value.isEmpty()) {
                    _errorMessage.value = error.message ?: "Failed to load events"
                }
            }

        _isLoading.value = false
        _isLoadingMore.value = false
    }

    /**
     * Fetch featured events. Mirrors iOS fetchFeaturedEvents().
     */
    private suspend fun fetchFeaturedEvents() {
        eventsRepository.fetchFeaturedEvents()
            .onSuccess { featured ->
                _featuredEvents.value = featured
            }
            .onFailure { error ->
                Log.e(TAG, "fetchFeaturedEvents failed: ${error.message}", error)
                // Keep existing cached featured events on failure
            }
    }

    /**
     * Fetch popular restaurants for the home screen carousel.
     * Uses default query sorted by popularity.
     */
    private suspend fun fetchPopularRestaurants() {
        restaurantsRepository.fetchRestaurants(
            RestaurantsQuery(
                sortBy = RestaurantSortOption.POPULARITY,
                isFeatured = true,
                limit = 10,
            )
        ).onSuccess { response ->
            _restaurants.value = response.restaurants
        }.onFailure { error ->
            Log.e(TAG, "fetchPopularRestaurants failed: ${error.message}", error)
        }
    }

    /**
     * Apply premium filters client-side. Mirrors iOS applyPremiumFilters(_:).
     */
    private fun applyPremiumFilters(events: List<Event>): List<Event> {
        var result = events

        if (_showFreeOnly.value) {
            result = result.filter { it.isFree }
        }

        val maxDist = _maxDistance.value
        if (maxDist != null) {
            val userLoc = locationService.userLocation.value
            val userLat = userLoc?.latitude ?: Config.DEFAULT_LATITUDE
            val userLng = userLoc?.longitude ?: Config.DEFAULT_LONGITUDE
            result = result.filter { event ->
                val coord = event.coordinate ?: return@filter true
                val distanceMiles = locationService.haversineDistanceMiles(
                    userLat, userLng, coord.latitude, coord.longitude
                )
                distanceMiles <= maxDist
            }
        }

        // minRating available for future use when events have ratings
        return result
    }

    /**
     * Debounced reset and fetch. Mirrors iOS resetAndFetch().
     * Cancels any pending fetch and waits 300ms before fetching.
     */
    private fun resetAndFetch() {
        debounceJob?.cancel()
        debounceJob = viewModelScope.launch {
            delay(Config.SEARCH_DEBOUNCE_MS)
            fetchEvents(reset = true)
        }
    }

    private fun computeActiveFilterCount(filters: FilterState): Int {
        var count = 0
        if (filters.selectedCategory != null) count++
        if (filters.selectedDatePreset != null) count++
        if (filters.showFeaturedOnly) count++
        if (filters.searchText.isNotEmpty()) count++
        if (filters.showFreeOnly) count++
        if (filters.maxDistance != null) count++
        if (filters.minRating != null) count++
        return count
    }

    // endregion
}

/**
 * Internal filter state holder for combine operations.
 */
private data class FilterState(
    val selectedCategory: EventCategory? = null,
    val selectedDatePreset: DateFilterPreset? = null,
    val showFeaturedOnly: Boolean = false,
    val searchText: String = "",
    val showFreeOnly: Boolean = false,
    val maxDistance: Double? = null,
    val minRating: Double? = null,
    val currentTier: SubscriptionTier = SubscriptionTier.FREE,
    val isConnected: Boolean = true,
)
